inngest-core/src/main/kotlin/com/inngest/signingkey/SignatureVerification.kt

```diff
@@ -0,0 +1,104 @@
+package com.inngest.signingkey
+
+import com.inngest.InngestEnv
+import com.inngest.ServeConfig
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+import java.time.Instant
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+
+const val HMAC_SHA256 = "HmacSHA256"
+
+// Implementation of this is inspired by the pure Java example from https://www.baeldung.com/java-hmac#hmac-using-jdk-apis
+@OptIn(ExperimentalStdlibApi::class)
+private fun computeHMAC(
+    algorithm: String,
+    data: String,
+    key: String,
+): String {
+    val secretKeySpec = SecretKeySpec(key.toByteArray(Charsets.UTF_8), algorithm)
+    val mac = Mac.getInstance(algorithm)
+    mac.init(secretKeySpec)
+    return mac.doFinal(data.toByteArray(Charsets.UTF_8)).toHexString()
+}
+
+internal fun signRequest(
+    requestBody: String,
+    timestamp: Long,
+    signingKey: String,
+): String {
+    return signRequest(requestBody, timestamp.toString(), signingKey)
+}
+
+private fun signRequest(
+    requestBody: String,
+    timestamp: String,
+    signingKey: String,
+): String {
+    val matchResult = SIGNING_KEY_REGEX.matchEntire(signingKey) ?: throw InvalidSigningKeyException()
+    val key = matchResult.groups["key"]!!.value
+    val message = requestBody + timestamp
+
+    return computeHMAC(HMAC_SHA256, message, key)
+}
+
+class InvalidSignatureHeaderException(message: String) : Throwable(message)
+
+class ExpiredSignatureHeaderException : Throwable("signature header has expired")
+
+const val FIVE_MINUTES_IN_SECONDS = 5L * 60
+
+internal fun validateSignature(
+    signatureHeader: String,
+    signingKey: String,
+    requestBody: String,
+) {
+    // TODO: Find a way to parse signatureHeader as URL params without constructing a full URL
+    val dummyUrl = "https://test.inngest.com/?$signatureHeader"
+    val url = dummyUrl.toHttpUrlOrNull() ?: throw InvalidSignatureHeaderException("signature header does not match expected format")
```

Comment:
this is janky but I wanted to stick to pure Java/Kotlin or using libraries we already have. In this case, we were already using okhttp.

The pure java examples were all raw string operations https://www.baeldung.com/kotlin/parsing-url-string, so I looked at the examples here and went with one that was fairly short and from a library already in use https://stackoverflow.com/a/48594445. Unfortunately it means we have to prefix with a dummy domain

Reply:
huh, that's interesting.

Reply:
yeah this might be overkill, based on https://github.com/inngest/inngest/blob/main/docs/SDK_SPEC.md#411-definitions
```
The format of the signature is a query string (without the leading ?) with two values: t and s, e.g. t=1705586504&s=signkey-prod-02jd7ak2. t is the timestamp for when the request was prepared by Inngest, ensuring old-but-valid requests cannot be repeated. s is the actual signature, which is an HMAC with SHA256 of the body of the request.
```
if we never have to worry about html encoded characters because it's not really a query string, then it should be a lot simpler to split on `&` and `=`. Also wasn't sure if the string always contained `t` and `s` in the same order.